File: src/main/java/model/players/AbstractPlayer.java
package model.players;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import model.Move;
import model.Tile;

public abstract class AbstractPlayer implements Player {

    private final List<Tile> tiles;
    private final String name;
    private int score;

    public AbstractPlayer(String _name) {
        this.name = _name;
        this.score = 0;
		this.tiles = new ArrayList<>();
    }

    public void resetTiles() {
        tiles.clear();
    }

    public void resetScore() {
        this.score = 0;
    }

    public String getName() {
        return name;
    }

    public int getScore() {
        return score;
    }

    public void addScore(int score) {
        this.score += score;
    }

    public List<Tile> getTiles() {
        return new ArrayList<>(tiles);
    }

    public void addTile(Tile tile) {
        this.tiles.add(tile);
    }

    public void removeTile(Tile tile) {
        this.tiles.remove(tile);
    }

    public Optional<Move> play(LinkedList<Tile> board) {
        return board.isEmpty() ? Optional.of(this.selectTileEmptyBoard())
                : this.hasPossibleTiles(board) ? Optional.of(this.selectTileNonEmptyBoard(board)) : Optional.empty();
    }

    public boolean hasPossibleTiles(LinkedList<Tile> board) {
        if (board.isEmpty()) {
            return true;
        }
        int first = board.peekFirst().getFirstNumber();
        int last = board.peekLast().getSecondNumber();
        return tiles.stream().anyMatch(t -> t.hasNumber(first) || t.hasNumber(last));
    }

    protected List<Tile> getPossibleTiles(int first, int last) {
        return tiles.stream().filter(t -> t.hasNumber(first) || t.hasNumber(last)).collect(Collectors.toList());
    }

    protected abstract Move selectTileEmptyBoard();

    protected abstract Move selectTileNonEmptyBoard(LinkedList<Tile> board);

    protected abstract String description();

}

File: src/main/webapp/scripts/players/AbstractPlayer.js
export default class AbstractPlayer {

    constructor(_name) {
        this.name = _name;
        this.score = 0;
		this.tiles = [];
    }

    resetTiles() {
        tiles = [];
    }

    resetScore() {
        this.score = 0;
    }

    getName() {
        return this.name;
    }

    getScore() {
        return this.score;
    }

    addScore(score) {
        this.score += score;
    }

    getTiles() {
        return [...tiles];
    }

    addTile(tile) {
        this.tiles.push(tile);
    }

    removeTile(tile) {
        this.tiles = this.tiles.filter(t => !t.equals(tile));
    }

    play(board) {
        return board.length === 0 ? this.selectTileEmptyBoard()
                : this.hasPossibleTiles(board) ? this.selectTileNonEmptyBoard(board) : [];
    }

    hasPossibleTiles(board) {
        if (board.length === 0) {
            return true;
        }
        let first = board[0].getFirstNumber();
        let last = board.slice(-1).getSecondNumber();
        return tiles.some(t => t.hasNumber(first) || t.hasNumber(last));
    }

    getPossibleTiles(first, last) {
        return tiles.filter(t => t.hasNumber(first) || t.hasNumber(last));
    }
}
